Add a TryGetLaserHitPosition method to UxrLaserPointer

XianGrim opened the pull request:
Uxr has a ton of methods for handling the laser pointer, but there isn't a method for getting the actual position of where the laser is hitting. This is what I came up with. I also tried to keep in line with UXR's coding style.

entromp commented:
Hello @XianGrim,

Thanks for contributing! I'm glad you mention this because thanks to your PR I believe I found some missing/misleading info in the laser pointer raycasting.

You're right that UxrLaserPointer has no direct access to the raycast hit world position. However, you can get it using:

```
UxrPointerEventData eventData = UxrPointerInputModule.Instance.GetLaserPointerEventData(laserPointer);

if (eventData.HasData)
{
    Debug.Log("WorldPos: " + eventData.pointerCurrentRaycast.worldPosition);
}
```

The UxrPointerEventData class has also a WorldPos, but I see that it's currently only used internally for finger tips to track the previous and current position and know if there was a touch in between. I think this should be clarified and also correctly filled with data when using laser pointers.

So, in addition to your method, I would suggest adding an additional TryGetPointerEventData method that would return the eventData in the code above. This would make it easier to access other data such as the object that was hit, etc.

What do you think?

XianGrim commented:
Sounds great to me. Originally I was using the method you wrote above but I had an issue where, after the first time of getting the position, any subsequent attempts would always return Vector3.zero. So, in my use case, when I moved a Unit to the laser's position, it worked the first time. But when trying to move it again, it would try to travel to 0,0,0.

Only using the method I wrote above was I able to get consistent results. Perhaps that has something to do with the missing data you mentioned?

entromp commented:
I just added the following fixes/improvements:
- UxrPointerEventData.WorldPos now is also updated for UxrLaserPointer components. Thanks for the heads up! I've modified your TryGetLaserHitPosition() to use this WorldPos internally instead.
- TryGetPointerEventData() has been added to both UxrLaserPointer and UxrFingerTip for quick access.

Thanks again for your contribution. If you're okay with the changes I will merge them into main and v1.0.0 :)

XianGrim commented:
> I just added the following fixes/improvements:
> 
>     * UxrPointerEventData.WorldPos now is also updated for UxrLaserPointer components. Thanks for the heads up! I've modified your TryGetLaserHitPosition() to use this WorldPos internally instead.
> 
>     * TryGetPointerEventData() has been added to both UxrLaserPointer and UxrFingerTip for quick access.
> 
> 
> Thanks again for your contribution. If you're okay with the changes I will merge them into main and v1.0.0 :)

Sounds good to me! Loving all the progress!
